index.js

```diff
@@ -9,46 +9,60 @@ class TimeoutError extends Error {
 	}
 }
 
-const pTimeout = (promise, milliseconds, fallback) => new Promise((resolve, reject) => {
-	if (typeof milliseconds !== 'number' || milliseconds < 0) {
-		throw new TypeError('Expected `milliseconds` to be a positive number');
-	}
-
-	if (milliseconds === Infinity) {
-		resolve(promise);
-		return;
-	}
+const pTimeout = (promise, milliseconds, fallback) => {
+	let timer;
 
-	const timer = setTimeout(() => {
-		if (typeof fallback === 'function') {
-			try {
-				resolve(fallback());
-			} catch (error) {
-				reject(error);
-			}
+	const p = new Promise((resolve, reject) => {
+		if (typeof milliseconds !== 'number' || milliseconds < 0) {
+			throw new TypeError('Expected `milliseconds` to be a positive number');
+		}
 
+		if (milliseconds === Infinity) {
+			resolve(promise);
 			return;
 		}
 
-		const message = typeof fallback === 'string' ? fallback : `Promise timed out after ${milliseconds} milliseconds`;
-		const timeoutError = fallback instanceof Error ? fallback : new TimeoutError(message);
+		timer = setTimeout(() => {
+			if (typeof fallback === 'function') {
+				try {
+					resolve(fallback());
+				} catch (error) {
+					reject(error);
+				}
 
-		if (typeof promise.cancel === 'function') {
-			promise.cancel();
-		}
+				return;
+			}
 
-		reject(timeoutError);
-	}, milliseconds);
+			const message = typeof fallback === 'string' ? fallback : `Promise timed out after ${milliseconds} milliseconds`;
+			const timeoutError = fallback instanceof Error ? fallback : new TimeoutError(message);
 
-	// TODO: Use native `finally` keyword when targeting Node.js 10
-	pFinally(
-		// eslint-disable-next-line promise/prefer-await-to-then
-		promise.then(resolve, reject),
-		() => {
-			clearTimeout(timer);
-		}
-	);
-});
+			if (typeof promise.cancel === 'function') {
+				promise.cancel();
+			}
+
+			reject(timeoutError);
+		}, milliseconds);
+
+		// TODO: Use native `finally` keyword when targeting Node.js 10
+		pFinally(
+			// eslint-disable-next-line promise/prefer-await-to-then
+			promise.then(resolve, reject),
+			() => {
+				clearTimeout(timer);
+			}
+		);
+	});
+
+	p.clear = () => {
+		p.canceled = true;
+		clearTimeout(timer);
+		timer = undefined;
+	};
+
+	p.canceled = false;
```

Comment:
Not sure about exposing this. Clearing the timeout is not the same as the promise being cancelled.